### PR TITLE
Stop header from jumping around

### LINKED ISSUE #175: Content jumping around while loading tracks
When searching/filtering etc, there's a lot of flashing happening caused by page elements jumping around:

https://www.loom.com/share/026b7c0909c84fd7a62e9fbf22547138

It feels like the whole page is being rerendered each time, causing everything to jump around, rather than just the parts that actually need to be re-rendered.

If only the relevant bits of content were changing the the page wouldn't be jumping and we wouldn't get this flashing effect.

We should fix this before doing more component stuff.

(cc @SleeplessByte for advice 🙂)

### PR DESCRIPTION
**Pull request by kntsoriano**
Closes https://github.com/exercism/v3-website/issues/175.

I've also fixed the tests in this PR.

## Preview

Oct-08-2020 13-53-19

### CODE REVIEW

File: app/javascript/components/student/TracksList.jsx [hunk 1]
```diff
@@ -32,10 +31,12 @@ function reducer(state, action) {
...
-  const { data, isSuccess, isError, isLoading } = useRequestQuery(
-    'track-list',
-    request
-  )
+  const {
+    resolvedData,
+    latestData,
+    isError,
+    isFetching,
+  } = usePaginatedRequestQuery('track-list', request)
```
Comment (ErikSchierboom) [inline app/javascript/components/student/TracksList.jsx:39]: It used to be `useRequestQuery`. Is there paging involved now?

File: app/javascript/components/student/TracksList.jsx [hunk 2]
```diff
@@ -46,16 +47,15 @@ export function TracksList({ statusOptions, tagOptions, ...props }) {
         </div>
       </div>
       <div className="md-container">
-        {isSuccess && <Header data={data} query={request.query} />}
+        <Header latestData={latestData} query={request.query} />
```
Comment (ErikSchierboom) [inline app/javascript/components/student/TracksList.jsx]: Ah, so this is the trick where we don't conditionally render the element but instead use the latest data. Nice!

Are there more places where we should be doing this?

File: app/javascript/components/student/tracks-list/Header.jsx [hunk 3]
```diff
@@ -1,14 +1,28 @@
-import React from 'react'
+import React, { useEffect, useState } from 'react'
 import pluralize from 'pluralize'
+import { Loading } from '../../common/Loading'
 
-export function Header({ data, query = {} }) {
+export function Header({ latestData, query = {} }) {
+  const [text, setText] = useState()
```
Comment (ErikSchierboom) [inline app/javascript/components/student/tracks-list/Header.jsx]: I don't know if this makes sense, but would initializing this with a default value also help reduce a re-render?

```suggestion
  const [text, setText] = useState("Exercism's Language Tracks")
```

File: test/javascript/components/student/TracksList.test.js [hunk 4]
```diff
@@ -12,9 +14,41 @@ test('disables the selected status filter', () => {
   const { getByText } = render(
     <TracksList
       statusOptions={options}
+      tagOptions={[]}
```
Comment (ErikSchierboom) [inline test/javascript/components/student/TracksList.test.js]: We'll be able to catch these errors once we've converted everything to TypeScript!

**Review by ErikSchierboom — APPROVED** [review]

File: app/javascript/components/student/tracks-list/Header.jsx (hunk 3 above)
Comment (iHiD) [reply app/javascript/components/student/tracks-list/Header.jsx]: This is a good idea I think.

File: app/javascript/components/student/tracks-list/Header.jsx [hunk 5]
```diff
@@ -1,14 +1,28 @@
-import React from 'react'
+import React, { useEffect, useState } from 'react'
 import pluralize from 'pluralize'
+import { Loading } from '../../common/Loading'
 
-export function Header({ data, query = {} }) {
+export function Header({ latestData, query = {} }) {
```
Comment (iHiD) [inline app/javascript/components/student/tracks-list/Header.jsx]: I don't get the reason for this rename inside this component. As far as the Header is concerned is this not always it's `data`?

**Review by iHiD — APPROVED** [review]
@kntsoriano I rebased this. I'll let you reply to the comments or update things as you see fit, and then merge

File: app/javascript/components/student/tracks-list/Header.jsx (hunk 5 above)
Comment (kntsoriano) [reply app/javascript/components/student/tracks-list/Header.jsx]: Ah yes. I'll rename it back to data.

File: app/javascript/components/student/tracks-list/Header.jsx (hunk 3 above)
Comment (kntsoriano) [reply app/javascript/components/student/tracks-list/Header.jsx]: Yeah. Good idea!

File: app/javascript/components/student/TracksList.jsx (hunk 2 above)
Comment (kntsoriano) [reply app/javascript/components/student/TracksList.jsx]: I think we will need to adapt this pattern more and more in the future. But let's leave that for later I think.

**Pull request merged**